censustree/gravitontree/trie.go

```diff
@@ -228,11 +243,15 @@ func (t *Tree) DumpPlain(root []byte) ([][]byte, [][]byte, error) {
 }
 
 // ImportDump imports a partial or whole tree previously exported with Dump()
-func (t *Tree) ImportDump(keys [][]byte) error {
+func (t *Tree) ImportDump(data []byte) error {
 	t.updateAccessTime()
 	var err error
-	for _, cb := range keys {
-		if err = t.Tree.Add(cb, []byte{}); err != nil {
+	census := new(exportData)
+	if err = bare.Unmarshal(data, census); err != nil {
```

Comment:
try to use `:=` in all these `if err := f(); err != nil {` pieces of code, then you often don't need the `var err error` at the top at all. plus, you don't have to worry about overwriting a previous error.

Reply:
Mmm, I'd like to understand what would be, in general, the best approach. 

To me declaring a single `err` variable for a whole function makes sense since it reuses a variable which is constantly used for the same objective (check an error immediately after receiving it). There is no possible overwriting because the check is always performed on the next line. 

What's the point for declaring new N error variables in a single function?

Reply:
From a human's point of view, variables should have the smallest scope possible. If a function shares an err variable for its entire body, there's a chance there will be a mistake somewhere, like:

```
var err error
// many lines here...
if err2 := f(); err2 != nil {
    return err // wrong one!
}
```

It's true that more variables are generally worse, but think of it from the point of view of variables in the *function body scope*. With your method, there's one more variable in that scope, not fewer.

Reply:
Also, more short-lived variables are not a problem for the compiler at all. It can be clever enough to just reuse registers or stack space for them.

Reply:
> From a human's point of view, variables should have the smallest scope possible. If a function shares an err variable for its entire body, there's a chance there will be a mistake somewhere, like:

Sure, it makes sense for most of the variables. But the err variable is somehow special in GoLang. Everyone is using just `err` most of the times. The example you posted is quite "weird" IMO, most of the users (even not experienced) would do:
```
var err error
// many lines here...
if err := f(); err != nil {
    return err // wrong one!
}
```

> Also, more short-lived variables are not a problem for the compiler at all. It can be clever enough to just reuse registers or stack space for them.

I'm pretty sure the GoLang compiler is quite smart and there is also the Garbage collector. However from the developer point of view, making things easier for the compiler/collector looks like a good approach.

In any case I don't have any very strong point for declaring a single `err` variable on functions. It's just I don't see any strong point either for not doing it... But if declaring a new err variable on each `if` statement is more idiomatic, lets follow this approach.

Btw it would be nice to have a document on go-dvote Wiki explaining best practices for writing code in our repositories.

Reply:
To be fair, some people do prefer the whole "declare err at the beginning of the func and reuse it". I don't think it's the most common style, so I wouldn't say it's "idiomatic". I think most people who use it are either used to old C, or aren't comfortable enough with variable shadowing and prefer to have a single declaration of err.

Regarding the doc; I think https://golang.org/doc/effective_go.html covers the basics. The "control structures" part touches a bit on `if err := ...`. I don't think we need a page of our own to describe how to write Go, except perhaps linking to existing well-maintained docs like that one.